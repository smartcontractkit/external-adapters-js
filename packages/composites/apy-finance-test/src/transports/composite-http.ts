import { EndpointContext } from '@chainlink/external-adapter-framework/adapter'
import { calculateHttpRequestKey } from '@chainlink/external-adapter-framework/cache'
import {
  TransportDependencies,
  TransportGenerics,
} from '@chainlink/external-adapter-framework/transports'
import { metrics } from '@chainlink/external-adapter-framework/metrics'
import { SubscriptionTransport } from '@chainlink/external-adapter-framework/transports/abstract/subscription'
import {
  TimestampedProviderResult,
  censorLogs,
  makeLogger,
  sleep,
} from '@chainlink/external-adapter-framework/util'
import { Requester } from '@chainlink/external-adapter-framework/util/requester'
import {
  AdapterDataProviderError,
  AdapterRateLimitError,
} from '@chainlink/external-adapter-framework/validation/error'
import { TypeFromDefinition } from '@chainlink/external-adapter-framework/validation/input-params'
import { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios'

const logger = makeLogger('CompositeHttpTransport')

export interface CompositeHttpTransportConfig<T extends TransportGenerics> {
  /**
   * This method should take the valid input parameters for a request, and make the data provider requests neccessary
   * to return the information.
   *
   *
   * @param params - valid input paramters sent to this Adapter
   * @param adapterSettings - the config for this Adapter
   * @param requestHandler - Function to make requests via Requester, enforcing rate limiting
   * @returns Promise that returns the timestamped result
   */
  performRequest: (
    params: TypeFromDefinition<T['Parameters']>,
    adapterSettings: T['Settings'],
    requestHandler: <R>(requestConfig: AxiosRequestConfig) => Promise<AxiosResponse<R>>,
  ) => Promise<TimestampedProviderResult<T>>
}

/**
 * Transport implementation that takes incoming batches of requests and keeps a warm cache of values.
 * Within the setup function, adapter params are added to a set that also keeps track and expires values.
 * In the background execute, the list of non-expired items in the set is fetched.
 * The list is then passed through to the `performRequest` function, that returns a Provider result.
 * This result is then set in the Cache so the adapter can fetch values from there.
 */
export class CompositeHttpTransport<T extends TransportGenerics> extends SubscriptionTransport<T> {
  // Flag used to track whether the warmer has moved from having no entries to having some and vice versa
  // Used for recording the cache warmer active metrics accurately
  WARMER_ACTIVE = false
  requester!: Requester

  constructor(private config: CompositeHttpTransportConfig<T>) {
    super()
  }

  override async initialize(
    dependencies: TransportDependencies<T>,
    adapterSettings: T['Settings'],
    endpointName: string,
    transportName: string,
  ): Promise<void> {
    await super.initialize(dependencies, adapterSettings, endpointName, transportName)
    this.requester = dependencies.requester
  }

  getSubscriptionTtlFromConfig(adapterSettings: T['Settings']): number {
    return adapterSettings.WARMUP_SUBSCRIPTION_TTL
  }

  async backgroundHandler(
    context: EndpointContext<T>,
    entries: TypeFromDefinition<T['Parameters']>[],
  ): Promise<void> {
    if (!entries.length) {
      logger.debug(
        `No entries in subscription set, sleeping for ${context.adapterSettings.BACKGROUND_EXECUTE_MS_HTTP}ms...`,
      )
      if (this.WARMER_ACTIVE) {
        // Decrement count when warmer changed from having entries to having none
        metrics.get('cacheWarmerCount').labels({ isBatched: 'true' }).dec()
        this.WARMER_ACTIVE = false
      }
      await sleep(context.adapterSettings.BACKGROUND_EXECUTE_MS_HTTP)
      return
    } else if (this.WARMER_ACTIVE === false) {
      // Increment count when warmer changed from having no entries to having some
      metrics.get('cacheWarmerCount').labels({ isBatched: 'true' }).inc()
      this.WARMER_ACTIVE = true
    }

    // We're awaiting these promises because although we have request coalescing, new entries
    // could be added to the subscription set if not blocking this operation, so the next time the
    // background execute is triggered if the request is for a fully batched endpoint, we could end up
    // with the full combination of possible params within the request queue
    logger.trace(`Handling ${entries.length} entries...`)
    const start = Date.now()
    await Promise.all(entries.map((e) => this.handleRequest(e, context)))
    const duration = Date.now() - start
    logger.trace(`All requests in the background execute were completed`)

    // These logs will surface warnings that operators should take action on, in case the execution of all
    // requests is taking too long so that entries could have expired within this timeframe
    if (duration > context.adapterSettings.WARMUP_SUBSCRIPTION_TTL) {
      logger.warn(
        `Background execution of all HTTP requests in a batch took ${duration},\
         which is longer than the subscription TTL (${context.adapterSettings.WARMUP_SUBSCRIPTION_TTL}).\
         This might be due to insufficient speed on the selected API tier, please check metrics and logs to confirm and consider moving to a faster tier.`,
      )
    }
    if (duration > context.adapterSettings.CACHE_MAX_AGE) {
      logger.warn(
        `Background execution of all HTTP requests in a batch took ${duration},\
         which is longer than the max cache age (${context.adapterSettings.CACHE_MAX_AGE}).\
         This might be due to insufficient speed on the selected API tier, please check metrics and logs to confirm and consider moving to a faster tier.`,
      )
    }

    // We're not sleeping here on purpose. We sleep when there are no entries in the subscription set to avoid polling too
    // frequently, but if we have entries we want the background execute to be re-run ASAP so we can prepare the next batch
    // of requests, and the sleep to rate-limit will be performed by the rate-limiter in the Requester.
    return
  }

  private async handleRequest(
    params: TypeFromDefinition<T['Parameters']>,
    context: EndpointContext<T>,
  ): Promise<void> {
    const { result, msUntilNextExecution } = await this.makeRequest(params, context)

    logger.debug('Setting adapter responses in cache')
    await this.responseCache.write(this.name, [result])

    if (msUntilNextExecution) {
      // If we got this, it means that the queue was unable to accomomdate this request.
      // We want to sleep here for a bit, to avoid running into constant queue overflow replacements in competing threads.
      logger.info(
        `Request queue has overflowed, sleeping for ${msUntilNextExecution}ms until reprocessing...`,
      )
      await sleep(msUntilNextExecution)
    }
  }

  private async makeRequest(
    params: TypeFromDefinition<T['Parameters']>,
    context: EndpointContext<T>,
  ): Promise<{
    result: TimestampedProviderResult<T>
    msUntilNextExecution?: number
  }> {
    const requestHandler = async <R>(
      requestConfig: AxiosRequestConfig,
    ): Promise<AxiosResponse<R>> => {
      const key = calculateHttpRequestKey({
        context: context,
        data: { ...(requestConfig.params || {}), ...(requestConfig.data || {}) },
        transportName: this.name,
      })

      const { response } = await this.requester.request<R>(key, requestConfig)
      return response
    }

    try {
      const result = await this.config.performRequest(
        params,
        context.adapterSettings,
        requestHandler,
      )
      return { result }
    } catch (e) {
      if (e instanceof AdapterDataProviderError) {
        const err = e as AdapterDataProviderError

        const errorMessage =
          err.cause instanceof AxiosError && err.cause.status
            ? `Data Provider request failed with status ${err.cause.status}: "${JSON.stringify(
                err.cause.response?.data,
              )}"`
            : `Data Provider request failed with error: ${err.cause}`

        censorLogs(() => logger.info(errorMessage))
        return {
          result: {
            params: params,
            response: {
              errorMessage,
              statusCode: 502,
              timestamps: err.timestamps,
            },
          },
        }
      } else if (e instanceof AdapterRateLimitError) {
        const err = e as AdapterRateLimitError
        censorLogs(() => logger.info(err.message))
        return {
          result: {
            params: params,
            response: {
              errorMessage: err.message,
              statusCode: 429,
              timestamps: {
                providerDataRequestedUnixMs: 0,
                providerDataReceivedUnixMs: 0,
                providerIndicatedTimeUnixMs: undefined,
              },
            },
          },
          msUntilNextExecution: err.msUntilNextExecution,
        }
      } else {
        censorLogs(() => logger.error(e))
        return {
          result: {
            params,
            response: {
              errorMessage: 'Unknown error',
              statusCode: 500,
              timestamps: {
                providerDataRequestedUnixMs: 0,
                providerDataReceivedUnixMs: 0,
                providerIndicatedTimeUnixMs: undefined,
              },
            },
          },
        }
      }
    }
  }
}
