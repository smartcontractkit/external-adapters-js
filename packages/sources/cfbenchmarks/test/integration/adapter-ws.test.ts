import { mockWebSocketServer } from './fixtures'
import {
  TestAdapter,
  setEnvVariables,
  mockWebSocketProvider,
  MockWebsocketServer,
} from '@chainlink/external-adapter-framework/util/testing-utils'
import { WebSocketClassProvider } from '@chainlink/external-adapter-framework/transports'
import FakeTimers from '@sinonjs/fake-timers'

describe('websocket', () => {
  let mockWsServer: MockWebsocketServer | undefined
  let testAdapter: TestAdapter
  const wsEndpoint = 'ws://localhost:9090'
  let oldEnv: NodeJS.ProcessEnv
  const dataCrypto = {
    index: 'BRTI',
  }
  const dataCryptoOverride = {
    base: 'ABC',
    quote: 'XXX',
    overrides: {
      cfbenchmarks: {
        ABC: 'BRTI',
      },
    },
  }
  const dataLwba = {
    from: 'ETH',
    quote: 'USD',
    endpoint: 'cryptolwba',
  }
  const dataLwbaInvariantViolation = {
    from: 'BTC',
    quote: 'USD',
    endpoint: 'cryptolwba',
  }
  const dataLwbaOverride = {
    endpoint: 'cryptolwba',
    base: 'LINK',
    quote: 'XXX',
    overrides: {
      cfbenchmarks: {
        LINK: 'U_LINKUSD_CHA_RTI',
      },
    },
  }

  beforeAll(async () => {
    oldEnv = JSON.parse(JSON.stringify(process.env))
    process.env['CACHE_MAX_AGE'] = '5000'
    process.env['CACHE_POLLING_MAX_RETRIES'] = '0'
    process.env['WS_API_ENDPOINT'] = wsEndpoint
    process.env['API_USERNAME'] = 'fake-api-username'
    process.env['API_PASSWORD'] = 'fake-api-password'

    // Start mock web socket server
    mockWebSocketProvider(WebSocketClassProvider)
    mockWsServer = mockWebSocketServer(wsEndpoint)

    const adapter = (await import('./../../src')).adapter
    testAdapter = await TestAdapter.startWithMockedCache(adapter, {
      clock: FakeTimers.install(),
      testAdapter: {} as TestAdapter<never>,
    })

    // Send initial request to start background execute and wait for cache to be filled with results
    await testAdapter.request(dataCrypto)
    await testAdapter.request(dataCryptoOverride)
    await testAdapter.request(dataLwba)
    await testAdapter.request(dataLwbaOverride)
    await testAdapter.request(dataLwbaInvariantViolation)
    await testAdapter.waitForCache(4)
  })

  afterAll(async () => {
    setEnvVariables(oldEnv)
    mockWsServer?.close()
    testAdapter.clock?.uninstall()
    await testAdapter.api.close()
  })

  describe('crypto endpoint', () => {
    it('should return success', async () => {
      const response = await testAdapter.request(dataCrypto)
      expect(response.json()).toMatchSnapshot()
    })

    it('with override should return success', async () => {
      const response = await testAdapter.request(dataCryptoOverride)
      expect(response.statusCode).toEqual(200)
      expect(response.json()).toMatchSnapshot()
    })
  })

  describe('crypto input validation', () => {
    it('should return error (empty data)', async () => {
      const response = await testAdapter.request({})
      expect(response.statusCode).toEqual(400)
    })

    it('should return error (empty base)', async () => {
      const response = await testAdapter.request({ quote: 'BTC' })
      expect(response.statusCode).toEqual(400)
    })

    it('should return error (empty quote)', async () => {
      const response = await testAdapter.request({ base: 'ETH' })
      expect(response.statusCode).toEqual(400)
    })
  })

  describe('lwba endpoint', () => {
    it('should return success', async () => {
      const response = await testAdapter.request(dataLwba)
      expect(response.json()).toMatchSnapshot()
    })

    it('with override should return success', async () => {
      const response = await testAdapter.request(dataLwbaOverride)
      expect(response.statusCode).toEqual(200)
      expect(response.json()).toMatchSnapshot()
    })

    it('should return error (LWBA violation)', async () => {
      const response = await testAdapter.request(dataLwbaInvariantViolation)
      expect(response.statusCode).toEqual(500)
      expect(response.json()).toMatchSnapshot()
    })
  })

  describe('lwba input validation', () => {
    it('should return error (empty data)', async () => {
      const response = await testAdapter.request({})
      expect(response.statusCode).toEqual(400)
    })

    it('should return error (empty base)', async () => {
      const response = await testAdapter.request({ quote: 'BTC' })
      expect(response.statusCode).toEqual(400)
    })

    it('should return error (empty quote)', async () => {
      const response = await testAdapter.request({ base: 'ETH' })
      expect(response.statusCode).toEqual(400)
    })
  })
})
