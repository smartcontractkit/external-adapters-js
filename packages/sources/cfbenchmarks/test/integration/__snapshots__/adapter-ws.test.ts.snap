// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`websocket crypto endpoint should return success 1`] = `
{
  "data": {
    "result": 40067,
  },
  "result": 40067,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1018,
    "providerDataStreamEstablishedUnixMs": 1010,
    "providerIndicatedTimeUnixMs": 1645203822000,
  },
}
`;

exports[`websocket crypto endpoint with override should return success 1`] = `
{
  "data": {
    "result": 40067,
  },
  "result": 40067,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1018,
    "providerDataStreamEstablishedUnixMs": 1010,
    "providerIndicatedTimeUnixMs": 1645203822000,
  },
}
`;

exports[`websocket lwba endpoint should return success 1`] = `
{
  "data": {
    "ask": 1.1662,
    "bid": 1.1607,
    "mid": 1.1635,
  },
  "result": null,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1018,
    "providerDataStreamEstablishedUnixMs": 1010,
    "providerIndicatedTimeUnixMs": 1677876163000,
  },
}
`;

exports[`websocket lwba endpoint with override should return success 1`] = `
{
  "data": {
    "ask": 2651.01,
    "bid": 2652.26,
    "mid": 2651.63,
  },
  "result": null,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1018,
    "providerDataStreamEstablishedUnixMs": 1010,
    "providerIndicatedTimeUnixMs": 1677876163000,
  },
}
`;
